### PR TITLE
[PVR][Estuary] Guide window: Channel groups selector

### PR DESCRIPTION
**Pull request by ksooo**
Requested many times and for a long time in the forum, here it comes... channel groups selector for the Guide window:

screenshot001

Channel groups are displayed as a "band" on top of the epg grid. Channel groups are switched live when selecting a group in the "band". Thus, a user can switch very quickly between channel groups.

The channel group selector can be turned off. There are two new view types for the Guide window "horizontal channels, no group selector" and "vertical channels, no group selector", in addition to the two existing viewtypes, which now contain the group selector. The views without the selector look exactly the same as the views we had before this PR. 

Changes are runtime-tested on macOS and Linux, latest kodi master.

@xhaggi mind taking a look at the code changes changes.
@ronie are the skin changes okay?

### CODE REVIEW

**Review by ronie — APPROVED** [review]
very nice!

**Comment by xhaggi** [comment]
I'm sure if i like it. Does this really improve the group selection? If you stay on channel 15 of 30, somewhere in the middle, how do you easily reach this group selection control? We put the channel group selection into the options menu which can be easi

**Comment by xhaggi** [comment]
Updated my first comment and sorry it was the wrong button. Never write comments on a phone ;)

**Comment by ksooo** [comment]
> If you stay on channel 15 of 30, somewhere in the middle, how do you easily reach this group selection control?

Hit "1" on your remote, that selects first channel row, than hit "up" and here we go.

**Comment by ksooo** [comment]
> I'm not sure if i like it. 

 If you don't like it switch to one of the views without the selector.

**Comment by ksooo** [comment]
> We put the channel group selection into the options menu which can be easily reached with the menu key

 I have no menu key on my remote. And I have to scroll left for "ages" to get to the side blade. That sucks.

**Comment by xhaggi** [comment]
>Hit "1" on your remote, that selects first channel row, than hit "up" and here we go. Just two key presses max, no matter were you are in the grid.

Why not focus the options menu if you leave the grip on top and down? That would be nearly the same. My concern is the the difference in behavior. Next some users request the same for the channel list.

**Comment by ksooo** [comment]
Why not adding this also to the channel list? Seriously.

**Comment by ksooo** [comment]
I really do not like that user experience when changing channel groups using the side blade. First, as I explained, without a menu key it is pain to get there. Then, you have to select the current group and to press select, then, a dialog pops up, where you have to select the new group, then press select again. Then, the new group populated in the guide window, which is half hidden behind the side blade. Finally, I have to hit back to close the side blade and to get beck to the guide window. So ugly, so many clicks...

**Comment by xhaggi** [comment]
I don't have anything against it in general, but it would be nice if we could unify it in the corresponding windows. Now we have different ways in different windows. Therefore it would be nice if the control could be used on the other windows as well. We could replace the channel group button with this blade control and if a skin want to place it in the options menu it should be functional the same. What do you think of that?

**Comment by jjd-uk** [comment]
Just for info, on my current Sky box (and other pvr’s I’ve used) the method to get to the channel group selection bar no matter where you are in epg grid is via the back key. As someone who doesn’t use Kodi pvr much I’ve no idea if that would fit in with current Kodi  navigation methods within the guide.

**Comment by xhaggi** [comment]
We have a dedicated key "menu" to reach the options menus in windows. On a full featured remote this is no issue. For it's easy to switch the channel group, not used very often but it's really no issue.

**Comment by zag2me** [comment]
Big.... awesome.... thanks for this, I literally was going to request this on the forums today. I hate the sideblade group selection and the PR makes it SO MUCH easier. Brilliant! Looks very similar to the sky EPG now which has great usability. Example here

**Comment by ksooo** [comment]
> On a full featured remote this is no issue.

All people not having a "fully featured" remote are having an issue with it. It's almost unusable, then.

> Therefore it would be nice if the control could be used on the other windows as well. 

 I moved the implementation to `CGUIWindowPVRBase` now. :-)

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRBase.cpp]: This should be moved to the first condition `if (InitChannelGroup())` block in this method.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRBase.cpp]: what's the reason for calling `CGUIMediaWindow::OnAction(action)` first?

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRBase.h]: could we rename the method to `ChannelGroupsControlHasFocus`

**Comment by da-anda** [comment]
ok, so my SHIELD remote does neither have a menu button nur numeric input. If I now bug long enough on the forum, will I also get my custom channel group switching method implementation? I have a back button, so jjd's solution would work ;)

What about the focus "overflow" of the EPG grid? Up until now I was able to press up on channel 1 to get to the end of the list. Will pressing "up" on the channel selector also end up at the end of the channel list, or is the overflow feature gone in this view mode?

Has it been tested what happens when there are more channel groups that exceed the available width? It might be an idea to change this control into one that has the currently selected channel group always at the same spot and the options just slide left and right. Something like a coverflow view. And only the selected channel group would be clearly visible, while the others would be dimmed (or entirely hidden until the control is focused). Just an idea.

Btw, it might be an idea to instantly focus this group selector when pressing "0" on your remote.

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRBase.h]: same would be nice here `ChannelGroupsControlSelectActiveGroup`

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRBase.h]: and here `ChannelGroupsControlSetSelectedGroup`

**Review by xhaggi — CHANGES_REQUESTED** [review]

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRBase.cpp]: why do we catch UP/DOWN here?

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: this is need to first update the control and later switch to the group, right?

**Comment by ksooo** [comment]
Sorry, I'm sick about this.

**Comment by xhaggi** [comment]
Sick about discussing a feature you want to merge? Or sick about my review? Could you please explain it a bit more for us.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Yes.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: What's wrong with this? Isn't this just personal freedom in coding things?

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: We always have an active group, with one important exception, if the channel groups have not been loaded yet. The code is fine as it is. So, if no active group, no groups at all, not possible to fill the control.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: This will stay as it is. Personal freedom.

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.h]: No. I prefer method names to start  with a verb.

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.h]: No.

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.h]: No. ;-)

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Good point. This is a leftover. Thanks.

**Comment by ksooo** [comment]
Your review is okay.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Sure it's a matter of taste, but it's more readable in that case and you already used early returns in this method.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: ok than please remove the superfluous null check below

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (xhaggi) [reply xbmc/pvr/windows/GUIWindowPVRBase.h]: The introduced method names `SelectActiveChannelGroup` and `ActivateSelectedChannelGroup` aren't self-explained (context is not clear), so why not change it to something that helps understanding the code. My first idea was to put this into a separate class but adding `ChannelGroupsControl` would be enough.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Yep, thanks. Does not harm where it is now, but fits better in the first block.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Which one? Cannot see a superfluous null check.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: I will change it the other way round and remove the other early return.

File: xbmc/pvr/windows/GUIWindowPVRBase.h
Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.h]: I don't think I will change that.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Line 333 you check `activeGroup` again there

Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Ah, okay. Overlooked this one. Thanks.

File: xbmc/pvr/windows/GUIWindowPVRBase.cpp
Comment (xhaggi) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Why? this will introduces more deep nesting. Sorry I don't know why not return and that's it.

Comment (ksooo) [reply xbmc/pvr/windows/GUIWindowPVRBase.cpp]: Because I don't want it. There is absolutely no sense in this kind of discussion. Its my personal coding style and freedom of choice. There is no right and wrong. Please, stop this.

**Comment by ksooo** [comment]
I have a final offer to make (code- and function-wise):

screenshot003
screenshot004
screenshot005
screenshot006

**Comment by ksooo** [comment]
> If I now bug long enough on the forum, will I also get my custom channel group switching method implementation?

 Please stop this kind of comment, please. I'm not open to this kind of "humor".

> Will pressing "up" on the channel selector also end up at the end of the channel list, or is the overflow feature gone in this view mode?

 The latter. Technically not possible atm. "Up" in the group selector goes back to where you were when setting focus to the selector (keyboard: row 1, mouse: the last row that had the focus.)

> Has it been tested what happens when there are more channel groups than screen width? 

 Ofc. Was working fine with initial implementation. Does work fine with my "final offer".

> It might be an idea to change this control into one that has the currently selected channel group always at the same spot and the options just slide left and right.

 Done in the "final offer".

> Btw, it might be an idea to instantly focus this group selector when pressing "0" on your remote.

 '0' is already taken. It positions the grid to 'now'. Otherwise, a good idea.

**Comment by ksooo** [comment]
> I don't have anything against it in general, but it would be nice if we could unify it in the corresponding windows. 

> Now we have different ways in different windows. 

 Not really. The old way still works, exactly like it was before. There is only another new way added. I don't see this as a problm.

> Therefore it would be nice if the control could be used on the other windows as well. 

 Done. Implementation is now in the base class of all PVR windows. Doing the nessacery skin changes for Estuary for the channel window is out of scope of this PR, though.

> We could replace the channel group button with this blade control and if a skin want to place it in the options menu it should be functional the same. What do you think of that?

 Sounds good. This is for another day (PR).

**Comment by ksooo** [comment]
Heads up please. I'm going to merge this if I don't get a resaonable veto during the next 12 hours.

**Comment by xhaggi** [comment]
I know that most of the comments are minor stuff, but in most of these cases you are not willing to compromise and want to push through your preferences. We both work in this area, but if you only allow your preferences, you don't need my review next time. Just push and merge, much easier. You can count yourself lucky to have some other dev working in the same area, we don't have much areas where 2 or more devs are working together.

**Comment by da-anda** [comment]
@ksooo yeah, sorry for the "joke" - I realized later that I should have deleted this paragraph as well. I initially also had a comment that I'm more on xhaggi's side and would prefer a one-fits-all approach instead of adding more code that needs maitenance, but I dropped it to not piss you off (which I still did in the end due to the first paragraph that I didn't remove). Oh well, lesson learned. I'd still be in favor of less settings options and a one-fits-all approach though, but I don't want to block this PR in any way.

Also, you don't have to instantly implement every suggestion I make. I'm just giving ideas that are open for discussion. I'm not demanding anything.

I'm not a skinner and have no idea if it's possible, but it might look better if the "focused" channel group is always on the left side (above the channel name column) and not in the center of the screen, along with showing the background only when focused. But this could maybe be fixed/adjusted later by a skinner? Or if one of our skinners would like to give it a spin right away in order to merge it into this PR, that would ofc be even nicer. @ronie @phil65 or anyone else?

**Comment by ksooo** [comment]
> I know that most of the comments are minor stuff, but in most of these cases you are not willing to compromise

And that is not a bad thing.

 Come on, you know me, I'm always willing to change my code if the reviewer requests more substantial things than "put a return statement where I want it to be" or "name something like I want it to be named". If something is not a bug/design flaw and purely cosmetical I tend to be not happy if I have to spend time on it, yeah. If there is no true and false both "parties" have **equal** rights. That's important.

**Comment by ronie** [comment]
@da-anda yeah, i can look into that, but it should block this PR from getting merged.

i'd say get it in as-is. if others feel the need to fine tune, they can do so in a follow-up PR.

**Comment by ksooo** [comment]
> Oh well, lesson learned

 Same here. I will most probably not touch anything UI-raleted for the next months. Seriously.

**Comment by ksooo** [comment]
And finally, I will leave it up to somebody else to hit the button for this PR...

**Comment by ksooo** [comment]
> yeah, i can look into 

 @ronie I would really much appreciate if you could beautify this a bit. My skinning experiences are ... well .. limited. ;-)

**Pull request merged**
